// @azure/.arm-desktopvirtualization-4zkjMK6b/dist-esm/src/operations/appAttachPackageInfo.js
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
import { __asyncDelegator, __asyncGenerator, __asyncValues, __await } from "tslib";
import { setContinuationToken } from "../pagingHelper";
import * as coreClient from "@azure/core-client";
import * as Mappers from "../models/mappers";
import * as Parameters from "../models/parameters";
/// <reference lib="esnext.asynciterable" />
/** Class containing AppAttachPackageInfo operations. */
export class AppAttachPackageInfoImpl {
    /**
     * Initialize a new instance of the class AppAttachPackageInfo class.
     * @param client Reference to the service client
     */
    constructor(client) {
        this.client = client;
    }
    /**
     * Gets information from a package given the path to the package.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param hostPoolName The name of the host pool within the specified resource group
     * @param importPackageInfoRequest Object containing URI to package image and other optional properties
     * @param options The options parameters.
     */
    listImport(resourceGroupName, hostPoolName, importPackageInfoRequest, options) {
        const iter = this.importPagingAll(resourceGroupName, hostPoolName, importPackageInfoRequest, options);
        return {
            next() {
                return iter.next();
            },
            [Symbol.asyncIterator]() {
                return this;
            },
            byPage: (settings) => {
                if (settings === null || settings === void 0 ? void 0 : settings.maxPageSize) {
                    throw new Error("maxPageSize is not supported by this operation.");
                }
                return this.importPagingPage(resourceGroupName, hostPoolName, importPackageInfoRequest, options, settings);
            },
        };
    }
    importPagingPage(resourceGroupName, hostPoolName, importPackageInfoRequest, options, settings) {
        return __asyncGenerator(this, arguments, function* importPagingPage_1() {
            let result;
            let continuationToken = settings === null || settings === void 0 ? void 0 : settings.continuationToken;
            if (!continuationToken) {
                result = yield __await(this._import(resourceGroupName, hostPoolName, importPackageInfoRequest, options));
                let page = result.value || [];
                continuationToken = result.nextLink;
                setContinuationToken(page, continuationToken);
                yield yield __await(page);
            }
            while (continuationToken) {
                result = yield __await(this._importNext(resourceGroupName, hostPoolName, importPackageInfoRequest, continuationToken, options));
                continuationToken = result.nextLink;
                let page = result.value || [];
                setContinuationToken(page, continuationToken);
                yield yield __await(page);
            }
        });
    }
    importPagingAll(resourceGroupName, hostPoolName, importPackageInfoRequest, options) {
        return __asyncGenerator(this, arguments, function* importPagingAll_1() {
            var _a, e_1, _b, _c;
            try {
                for (var _d = true, _e = __asyncValues(this.importPagingPage(resourceGroupName, hostPoolName, importPackageInfoRequest, options)), _f; _f = yield __await(_e.next()), _a = _f.done, !_a; _d = true) {
                    _c = _f.value;
                    _d = false;
                    const page = _c;
                    yield __await(yield* __asyncDelegator(__asyncValues(page)));
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (!_d && !_a && (_b = _e.return)) yield __await(_b.call(_e));
                }
                finally { if (e_1) throw e_1.error; }
            }
        });
    }
    /**
     * Gets information from a package given the path to the package.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param hostPoolName The name of the host pool within the specified resource group
     * @param importPackageInfoRequest Object containing URI to package image and other optional properties
     * @param options The options parameters.
     */
    _import(resourceGroupName, hostPoolName, importPackageInfoRequest, options) {
        return this.client.sendOperationRequest({ resourceGroupName, hostPoolName, importPackageInfoRequest, options }, importOperationSpec);
    }
    /**
     * ImportNext
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param hostPoolName The name of the host pool within the specified resource group
     * @param importPackageInfoRequest Object containing URI to package image and other optional properties
     * @param nextLink The nextLink from the previous successful call to the Import method.
     * @param options The options parameters.
     */
    _importNext(resourceGroupName, hostPoolName, importPackageInfoRequest, nextLink, options) {
        return this.client.sendOperationRequest({
            resourceGroupName,
            hostPoolName,
            importPackageInfoRequest,
            nextLink,
            options,
        }, importNextOperationSpec);
    }
}
// Operation Specifications
const serializer = coreClient.createSerializer(Mappers, /* isXml */ false);
const importOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DesktopVirtualization/hostPools/{hostPoolName}/importAppAttachPackageInfo",
    httpMethod: "POST",
    responses: {
        200: {
            bodyMapper: Mappers.AppAttachPackageList,
        },
        default: {
            bodyMapper: Mappers.ErrorResponse,
        },
    },
    requestBody: Parameters.importPackageInfoRequest,
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.hostPoolName,
    ],
    headerParameters: [Parameters.accept, Parameters.contentType],
    mediaType: "json",
    serializer,
};
const importNextOperationSpec = {
    path: "{nextLink}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.AppAttachPackageList,
        },
        default: {
            bodyMapper: Mappers.ErrorResponse,
        },
    },
    urlParameters: [
        Parameters.$host,
        Parameters.nextLink,
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.hostPoolName,
    ],
    headerParameters: [Parameters.accept, Parameters.contentType],
    mediaType: "json",
    serializer,
};
